### PR TITLE
Allow indexing tensors with both CPU and CUDA tensors

### PR DESCRIPTION
**Pull request by zou3519**
This copies `indices` to the same device as `src` and then performs the indexing.

cc @colesbury 

### Test Plan
code reading
Unit test (it's a sanity test and probably doesn't hit any edge cases, but I'm not sure what those look like)

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import re
 import unittest
 from itertools import repeat
+import random
```
Comment (colesbury) [inline test/test_cuda.py]: is this import used?

**Comment by apaszke** [comment]
Do we want to add it to improve compatibility with CPU scalars? I don't think it's a good idea to support cross-device indexing with non-scalar tensors, since we generally disallow such operations (they are very expensive!).

**Comment by fmassa** [comment]
@apaszke I think this is fixing a regression, as I believe [we used to support indexing CUDA tensors with CPU tensors](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorIndex.cu#L95-L105)

**Comment by zou3519** [comment]
@apaszke the context is that right now the error message for indexing a CUDA tensor with a CPU tensor is wrong: https://github.com/pytorch/pytorch/issues/5559 and I thought it would be easiest to fix by allowing cross-device indexing. If we think this isn't a good idea I can go back and just fix the error message

**Comment by colesbury** [comment]
I think this is the right strategy. The indexing operations allow many different types to be used as indices. For example, you can index a CUDA tensor with a Python list or NumPy array. It seems to follow that you should be able to use a PyTorch CPU tensor as well.

**Comment by yf225** [comment]
@zou3519 seems that there is intermittent error on Windows: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test/2470//console

**Comment by zou3519** [comment]
@yf225 taking a look

**Pull request merged**